config_lib/parsers/parser_yaml.py

```diff
@@ -1,11 +1,225 @@
+        sub_indent = _get_line_indent(sub_line)
+        if sub_indent <= min_indent:
+            break
+        sub_lines.append(sub_line)
+        index += 1
+
+    return sub_lines, index
+
+
+def _parse_list_item_with_key_value(item_line, current_indent, lines, index,
+                                    base_index):
+    key_part, value_part = item_line.split(':', 1)
+    key = key_part.strip()
+    value = value_part.strip()
+    item_dict = {}
+    line_number = base_index + index + 1
+    current_item_indent = current_indent + 2
+
+    if not key:
+        raise YAMLSyntaxError("Empty key in list item", line_number)
+
+    if value:
+        item_dict[key] = _parse_yaml_scalar(value)
+    else:
+        sub_lines, new_index = _collect_nested_lines(lines, index + 1,
+                                                     current_indent)
+        if sub_lines:
+            item_dict[key] = _parse_yaml_lines(sub_lines, current_item_indent,
+                                               base_index + index + 1)
+        index = new_index - 1
+
+    sub_lines, new_index = _collect_nested_lines(lines, index + 1,
+                                                 current_indent)
+    if sub_lines:
+        nested_dict = _parse_yaml_lines(sub_lines, current_item_indent,
+                                        base_index + index + 1)
+        if isinstance(nested_dict, dict):
+            item_dict.update(nested_dict)
+
+    return item_dict, new_index - 1
+
+
+def _parse_list_item(line, current_indent, lines, index, base_index, result):
+    line_number = base_index + index + 1
+
+    if result is None:
+        result = []
+    elif not isinstance(result, list):
+        raise YAMLSyntaxError("Mixed list and dict structures are not allowed",
+                              line_number)
+
+    item_line = line.lstrip()[2:]
+    current_item_indent = current_indent + 2
+
+    if ':' in item_line:
+        item_dict, new_index = _parse_list_item_with_key_value(
+            item_line, current_indent, lines, index, base_index)
+        result.append(item_dict)
+        return result, new_index
+
+    elif item_line == '':
+        sub_lines, new_index = _collect_nested_lines(lines, index + 1,
+                                                     current_indent)
+        nested_value = _parse_yaml_lines(sub_lines, current_item_indent,
+                                         base_index + index + 1)
+        result.append(nested_value)
+        return result, new_index - 1
+
+    else:
+        result.append(_parse_yaml_scalar(item_line))
+        return result, index
+
+
+def _parse_key_value_pair(line, current_indent, lines, index, base_index,
+                          result):
+    line_number = base_index + index + 1
+
+    if result is None:
+        result = {}
+    elif not isinstance(result, dict):
+        raise YAMLSyntaxError("Mixed list and dict structures are not allowed",
+                              line_number)
```

Comment:
The elif block after return is unnecessary because once a return statement is executed, the function exits, and any code that follows will not be executed. It’s better to do it like this: 
```suggestion
    if result is None:
        result = {}
    if not isinstance(result, dict):
        raise YAMLSyntaxError("Mixed list and dict structures are not allowed",
                              line_number)
```